satriani/satriani.interpreter.js

```diff
@@ -169,8 +174,8 @@ function evaluate(tree, env) {
                 }
                 return;
             case "comparison":
-                let lhs = evaluate(expr.lhs, env);
-                let rhs = evaluate(expr.rhs, env);
+                let lhs = toScalar(evaluate(expr.lhs, env));
+                let rhs = toScalar(evaluate(expr.rhs, env));
```

Comment:
At a glance, it looks like this will compare arrays as equal if they're the same length... am I reading this right? And if so, does that make sense? I haven't given much consideration to array semantics, and comparison based on length would have the convenient side-effect that the empty array is equal to zero, but... I'm not 100% sure about it?

Reply:
Well spotted -- we should leave special treatment of arrays to the `eq` function, which already has a special case for that. I think this means I should add some tests to check array equality.

Reply:
While making this change, I found that the existing implementation of `eq_array` tries to compare elements inside an array to determine equality, but the [Binary Comparison section of the spec](https://github.com/RockstarLang/rockstar/blob/b4e1cf9075eb531d9739b909b603f6b200be094a/spec.md#binary-comparison) says "Objects are checked by reference equality; all other types are checked by value equality". I can't find any other notes on what should(n't) be classified as an object, though.

Should arrays be compared element-wise, or should their identities be compared?

If the former, should the elements be compared using their identities or using `eq` (which coerces to compare across types)? Should `[null]` and `[0]` be equal?

If the latter, does that mean that `let alias be original`/`put original in alias` would need to preserve array references? I'm not sure there's a way to copy arrays by reference in the current reference (😆) implementation.